frontend/catalyst/from_plxpr/from_plxpr.py

```diff
@@ -240,6 +261,28 @@ def calling_convention(*args):
 }
 
 
+def apply_compiler_decompose_to_plxpr(inner_jaxpr, consts, tgatesets, ncargs):
+    """Apply the compiler-specific decomposition for a given JAXPR."""
+
+    # disable the graph decomposition optimization
+    is_graph = qml.decomposition.enabled_graph()
+    if is_graph:
+        qml.decomposition.disable_graph()
+
+    # First perform the pre-mlir decomposition to simplify the jaxpr
+    # by decomposing high-level gates and templates
+    gate_set = COMPILER_OPERATIONS + list(set().union(*tgatesets))
```

Comment:
The `gate_set` should always be the device-native gate set. Even in the first stage decomposition, the graph should be constructed and solved towards the final device-native gate set, but your `stopping_condition` can be specified in terms of `COMPILER_OPERATIONS`. This way the same graph solution can be used again in the second stage.

Reply:
This was my original plan w/ re-using the current `DecomposeInterpreter` but here for Catalyst, we are dealing with a two-phased decomposition pipeline in `from_plxpr`:
1. compiler-specific decomposition that decomposes the program w/ `COMPILER_OPERATIONS`. We need this to transform the qnode prior to other transforms. That's why I've included this in `handle_qnode`.  
2. target gateset decomposition that walks through the program after the first pass, collects all operations, and builds and solves the graph. At this step we need to also add an MLIR handler to the compiled program that wraps the whole qnode (after compiler-specific decomp) in it which will be handled by a custom compiler-lowering registered pass in `handle_transforms`.

Reply:
if by lowering you mean lowering to MLIR, we don't lower any rules at this first stage. We create a decomposed closed jaxpr w.r..t. the compiler ops